<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU53_impacto de las políticas de inversión en sanidad, infraestructuras y promoción turística en el SPI			</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 02. Transformar datos del SPI

* Convertir a csv los datos del SPI, que vienen en libro Excel con indicadores en columnas







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Encoding

Con la siguiente expresión se evitan problemas con el encoding al ejecutar el notebook. Es posible que deba ser eliminada o adaptada a la máquina en la que se ejecute el código.

In [1]:
Sys.setlocale(category = "LC_ALL", locale = "es_ES.UTF-8")

[1] "es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/C"

### Packages to use

_ELIMINAR O AÑADIR LO QUE TOQUE. COPIAR VERSIONES AL FINAL Y QUITAR CÓDIGO DE VERSIONES_

- {tcltk} para selección interactiva de archivos locales
- {sf} para trabajar con georeferenciación
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {stringr} para manipulación de cadenas de caracteres
- {tidyr} para organización de datos


In [2]:
library(readxl)
library(readr)
library(dplyr)
library(stringr)
library(tidyr)

p <- c("tcltk", "readxl", "readr", "dplyr", "stringr", "tidyr")



Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




### Paths

In [3]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [4]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [5]:
iFile <- "2022_social_progress_index_dataset-1663250262.xlsx"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/2022_social_progress_index_dataset-1663250262.xlsx

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

* Se leen primero los metadatos de los índices

In [6]:
metadata_01 <- read_excel(file_data, 
                       sheet = "2011-2022 SPI data", col_names = FALSE, 
                       skip = 1, n_max = 2) |> 
  t() |> as.data.frame(row.names = FALSE, col.names = c("id_var", "name_var")) |> 
  rename(id_var = V1, name_var = V2)

New names:
• `` -> `...1`
• `` -> `...2`
• `` -> `...3`
• `` -> `...4`
• `` -> `...5`
• `` -> `...6`
• `` -> `...7`
• `` -> `...8`
• `` -> `...9`
• `` -> `...10`
• `` -> `...11`
• `` -> `...12`
• `` -> `...13`
• `` -> `...14`
• `` -> `...15`
• `` -> `...16`
• `` -> `...17`
• `` -> `...18`
• `` -> `...19`
• `` -> `...20`
• `` -> `...21`
• `` -> `...22`
• `` -> `...23`
• `` -> `...24`
• `` -> `...25`
• `` -> `...26`
• `` -> `...27`
• `` -> `...28`
• `` -> `...29`
• `` -> `...30`
• `` -> `...31`
• `` -> `...32`
• `` -> `...33`
• `` -> `...34`
• `` -> `...35`
• `` -> `...36`
• `` -> `...37`
• `` -> `...38`
• `` -> `...39`
• `` -> `...40`
• `` -> `...41`
• `` -> `...42`
• `` -> `...43`
• `` -> `...44`
• `` -> `...45`
• `` -> `...46`
• `` -> `...47`
• `` -> `...48`
• `` -> `...49`
• `` -> `...50`
• `` -> `...51`
• `` -> `...52`
• `` -> `...53`
• `` -> `...54`
• `` -> `...55`
• `` -> `...56`
• `` -> `...57`
• `` -> `...58`
• `` -> `...59`
• `` -> `...60`
• `` -> `...61`
• `` -> `...62`
• `` -

In [7]:
metadata_01 |> glimpse()

Rows: 81
Columns: 2
$ id_var   <chr> "rank_score_spi", "country", "spicountrycode", "spiyear", "st…
$ name_var <chr> "SPI \r\nRank", "Country", "SPI country code", "SPI \r\nyear"…


In [8]:
metadata_01 |> slice_head(n = 10)

id_var         name_var                      
1  rank_score_spi SPI \r\nRank                  
2  country        Country                       
3  spicountrycode SPI country code              
4  spiyear        SPI \r\nyear                  
5  status         Status                        
6  score_spi      Social Progress Index         
7  score_bhn      Basic Human Needs             
8  score_fow      Foundations of Wellbeing      
9  score_opp      Opportunity                   
10 score_nbmc     Nutrition & Basic Medical Care

* Ahora se leen los índices, con los nombres de columnas de los metadatos

In [9]:
data <- read_excel(file_data, 
                  sheet = "2011-2022 SPI data", 
                  col_names = metadata_01$id_var, 
                  skip = 3)


Estructura de  los datos:

In [10]:
data |> glimpse()

Rows: 2,364
Columns: 81
$ rank_score_spi       <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, 1…
$ country              <chr> "World", "World", "World", "World", "World", "Wor…
$ spicountrycode       <chr> "WWW", "WWW", "WWW", "WWW", "WWW", "WWW", "WWW", …
$ spiyear              <dbl> 2022, 2021, 2020, 2019, 2018, 2017, 2016, 2015, 2…
$ status               <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, "…
$ score_spi            <dbl> 65.24, 64.87, 64.55, 64.31, 63.62, 63.29, 62.65, …
$ score_bhn            <dbl> 75.80, 75.35, 74.91, 74.48, 73.98, 73.43, 72.88, …
$ score_fow            <dbl> 63.62, 63.18, 62.91, 62.46, 61.03, 60.52, 59.26, …
$ score_opp            <dbl> 56.28, 56.07, 55.83, 55.98, 55.84, 55.92, 55.80, …
$ score_nbmc           <dbl> 82.09, 82.01, 81.79, 81.53, 81.23, 80.90, 80.53, …
$ score_ws             <dbl> 80.74, 79.84, 78.69, 77.99, 77.20, 76.59, 76.17, …
$ score_sh             <dbl> 79.36, 78.86, 78.50, 77.63, 76.92, 75.92, 74.97, …
$ score_ps      

Muestra de datos:

In [11]:
data |> slice_head(n = 5)

rank_score_spi country spicountrycode spiyear status score_spi score_bhn
1 NA             World   WWW            2022    NA     65.24     75.80    
2 NA             World   WWW            2021    NA     64.87     75.35    
3 NA             World   WWW            2020    NA     64.55     74.91    
4 NA             World   WWW            2019    NA     64.31     74.48    
5 NA             World   WWW            2018    NA     63.62     73.98    
  score_fow score_opp score_nbmc ... incl_equalaccess incl_sexualorient
1 63.62     56.28     82.09      ... 0.5446           0.8111           
2 63.18     56.07     82.01      ... 0.5460           0.7890           
3 62.91     55.83     81.79      ... 0.5225           0.7612           
4 62.46     55.98     81.53      ... 0.5603           0.8102           
5 61.03     55.84     81.23      ... 0.5557           0.7837           
  incl_accpubsersocgr incl_gayslesb incl_discrimin aae_acadfreed
1 1.9767              0.4127        7.1472         0.4374       
2 1.9979              0.3851        7.3180         0.4784       
3 2.0361              0.3506        7.2646         0.4647       
4 2.0183              0.3064        7.4035         0.4733       
5 2.0772              0.2839        7.5010         0.4836       
  aae_femterteduc aae_tertschlife aae_citabledocs aae_qualuniversities
1 0.3450          2.1666          0.5862          217.6475            
2 0.3342          2.1218          0.5438          225.5760            
3 0.3235          2.0671          0.5146          211.4173            
4 0.3132          2.0268          0.4859          203.6117            
5 0.3032          2.0050          0.4609          185.5802

* Obtenemos una segunda parte de los metadatos

In [12]:
metadata_02 <- read_excel(file_data, 
                         sheet = "DEFINITIONS") |> 
  drop_na(`Indicator name`)

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Datos de índices
* Metadatos de índices (diccionario de datos)

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

Si no aplica: Estos datos no requieren tareas de este tipo.

#### Data transformation

* Transformar metadatos:
    - Eliminar saltos de línea
    - Separar unidades
    - Sustituir & por and
    - Unir dimensiones
    - Añadir indicador del rol del valor

In [13]:
tmetadata <- metadata_01 |> 
  mutate(name_var = str_remove_all(name_var, "\\r\\n")) |> 
  separate(name_var, into = c("name_var", "tmp_units"), sep = " \\(") |>  
  mutate(name_var = str_trim(name_var)) |> 
  mutate(name_var = str_replace_all(name_var, "\\&", "and")) |> 
  select(-tmp_units) |> 
  mutate(name_var = str_to_sentence(name_var)) |> 
  left_join(metadata_02 |> select(`Indicator name`, Dimension, 
                                 Component, `Unit of measurement`,
                                 Definition, Source), 
            by = c("name_var" = "Indicator name")) |> 
  mutate(role = c(rep("aux", 5), "SPI", 
                  rep("Dimension", 3),
                  rep("Component", 3*4),
                  rep("Indicator", 60)),
         .after = "name_var")

Warning message:
“Expected 2 pieces. Missing pieces filled with `NA` in 21 rows [1, 2, 3, 4, 5,
6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...].”


In [14]:
tmetadata |> glimpse()

Rows: 81
Columns: 8
$ id_var                <chr> "rank_score_spi", "country", "spicountrycode", "…
$ name_var              <chr> "Spi rank", "Country", "Spi country code", "Spi …
$ role                  <chr> "aux", "aux", "aux", "aux", "aux", "SPI", "Dimen…
$ Dimension             <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ Component             <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ `Unit of measurement` <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ Definition            <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ Source                <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …


In [15]:
tmetadata |> slice(c(1:3, 22:26))


id_var              name_var                role      Dimension        
1 rank_score_spi      Spi rank                aux       NA               
2 country             Country                 aux       NA               
3 spicountrycode      Spi country code        aux       NA               
4 nbmc_stunting       Child stunting          Indicator Basic Human Needs
5 nbmc_infectiousdaly Infectious diseases     Indicator Basic Human Needs
6 nbmc_matmort        Maternal mortality rate Indicator Basic Human Needs
7 nbmc_childmort      Child mortality rate    Indicator Basic Human Needs
8 nbmc_undernourish   Undernourishment        Indicator Basic Human Needs
  Component                        Unit of measurement         
1 NA                               NA                          
2 NA                               NA                          
3 NA                               NA                          
4 Nutrition and Basic Medical Care (0=low risk; 100=high risk) 
5 Nutrition and Basic Medical Care (DALYs/100,000)             
6 Nutrition and Basic Medical Care (deaths/100,000 live births)
7 Nutrition and Basic Medical Care (deaths/1,000 live births)  
8 Nutrition and Basic Medical Care (% of pop.)                 
  Definition                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
1 NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
2 NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
3 NA                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
4 Risk-weighted prevalence of stunting in children under 5 as measured by the su

## Synthetic Data Generation

No Aplica

## Fake Data Generation

No aplica


## Open Data

No aplica

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

1. Datos del SPI

<font color='tomato'> Identificamos los datos a guardar</font>

In [16]:
data_to_save_01 <- data


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [17]:
caso <- "CU_53"
proceso <- '_05'
tarea <- "_02"
archivo <- "_01"
proper <- "_spi"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [18]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save_01, path_out_01)

# cat('File saved as: ')
# path_out_01

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [19]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save_01, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_53_05_02_01_spi.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [20]:
path_in_01 <- paste0(iPath, file_save_01)
file.copy(path_out_01, path_in_01, overwrite = TRUE)

[1] TRUE

2. Metadatos del SPI

<font color='tomato'> Identificamos los datos a guardar</font>

In [22]:
data_to_save_02 <- tmetadata


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [23]:
archivo <- "_02"
proper <- "_spi_metadata"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [24]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out_02

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [25]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data_to_save_02, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_53_05_02_02_spi_metadata.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [26]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes readxl, readr, dplyr, stringr, tidyr deben estar instalados.










### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* readxl 1.4.1
* readr 2.1.3
* dplyr 1.0.10
* stringr 1.5.0
* tidyr 1.3.0

### Data structures

#### Objeto `data`

- Hay 2364 filas con información de las siguientes variables:
	* rank_score_spi
	* country
	* spicountrycode
	* spiyear
	* status
	* score_spi
	* score_bhn
	* score_fow
	* score_opp
	* score_nbmc
	* score_ws
	* score_sh
	* score_ps
	* score_abk
	* score_aic
	* score_hw
	* score_eq
	* score_pr
	* score_pfc
	* score_incl
	* score_aae
	* nbmc_stunting
	* nbmc_infectiousdaly
	* nbmc_matmort
	* nbmc_childmort
	* nbmc_undernourish
	* nbmc_dietlowfruitveg
	* ws_washmortdalys
	* ws_sanitation
	* ws_water
	* ws_watersat
	* sh_hhairpolldalys
	* sh_affhousingdissat
	* sh_electricity
	* sh_cleanfuels
	* ps_politicalkillings
	* ps_intpersvioldaly
	* ps_transportdaly
	* ps_intimpartnviol
	* ps_moneystolen
	* abk_qualeduc
	* abk_propnoeduc
	* abk_popsomesec
	* abk_totprimenrol
	* abk_educpar
	* aic_altinfo
	* aic_mobiles
	* aic_internet
	* aic_eparticip
	* hw_qualityhealth
	* hw_lifex60
	* hw_ncdmort
	* hw_univhealthcov
	* hw_qualhealthsat
	* eq_airpolldalys
	* eq_leadexpdalys
	* eq_pm25
	* eq_spindex
	* pr_freerelig
	* pr_proprightswomen
	* pr_peaceassemb
	* pr_accessjustice
	* pr_freediscuss
	* pr_polrights
	* pfc_freedomestmov
	* pfc_earlymarriage
	* pfc_contracept
	* pfc_neet
	* pfc_vulnemploy
	* pfc_corruption
	* incl_equalprotect
	* incl_equalaccess
	* incl_sexualorient
	* incl_accpubsersocgr
	* incl_gayslesb
	* incl_discrimin
	* aae_acadfreed
	* aae_femterteduc
	* aae_tertschlife
	* aae_citabledocs
	* aae_qualuniversities


#### Objeto `tmetadata`

- Hay 81 filas con información de las siguientes variables:
	* id_var
	* name_var
	* role
	* Dimension
	* Component
	* Unit of measurement
	* Definition
	* Source

#### Observaciones generales sobre los datos

* Los datos son jerárquicos: los indicadores forman componentes, los componentes dimensiones y las dimensiones el índice.

### Consideraciones para despliegue en piloto

* Los metadatos son útiles para la representación en tablas y gráficos con etiqueteas amigables

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados
* Los datos de entrada se deben adquirir para usar en producción

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han dispuesto los índices en formato rectangular con nombres de columna adecuados

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben relacionar estos datos con los de inversiones

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código